### DISCUSSION TITLE
How does the Jupyter extension control the Notebook File Root?

### DISCUSSION THREAD

**Discussion by itcarroll** [microsoft/vscode-jupyter#13994] [Tips, Tricks and Workarounds]
The default (for me) setting of Notebook File Root (jupyter.notebookFileRoot) is ${workspaceDir}, and this is a really nice improvement over JupyterLab. The JupyterLab project doesn't leverage its "workspace" capability to function as an IDE that is aware of the project's root, so JupyterLab uses the equivalent of ${fileDirname} for the analagous setting. I would like to port this VSCode Jupyter extension behavior to a JupyterLab configuration ... but can't figure out any means to do so!? How does the Jupyter extension achieve this when launching the kernel? Pointer to the relevant code, please?

**Comment by DonJayamanne** [comment]
This happens after we start the kernel, 
once the kernel has been started we execute some python code. Sadly this feature only works for Python kernels, hence its not a generic solution.

Have a look at calculateWorkingDirectory in https://github.com/microsoft/vscode-jupyter/blob/main/src/interactive-window/kernelStartupCodeProvider.node.ts

The `getCode` method in the above file is called in `gatherInternalStartupCode` in the https://github.com/microsoft/vscode-jupyter/blob/main/src/kernels/kernel.ts file
And you can see that getting executed here in the same file (just after the kernel has been started)
```typescript
            const startupCode = await this.gatherInternalStartupCode();
            await this.executeSilently(session, startupCode, {
                traceErrors: true,
                traceErrorsMessage: 'Error executing jupyter extension internal startup code'
            });
```

**Reply by itcarroll** [reply]
Got it, thank you! I guess the fact that you have to rely on executing some Python code indicates I am not alone in finding no way to configure a kernel spec to do this for you. Harumph!
